### PR TITLE
Fixes #3096 - propagate time dependent changes

### LINKED ISSUE #3096: Align Action methods with coupler methods
The task went a bit sideways so the original title and description are wrong.   The work done is associated with the original title, but the specific actions were handled in a separate PR.

### PR DESCRIPTION
**Pull request by tclune**
This commit provides low level capabilities for computing the difference in metadata (units, typekind, geom, ...) between 2 fields or field bundles as well as applying diffs to a targeted field/bundle.

The intent is for couplers to use this to propagate time-varying attributes between import and export.   Changes will generally flow in each direction.

Intermediate progress.

May have to start over splitting into finer tasks.

Update/reallocate on fields moved

- New class FieldDelta and modified tests now in Test_FieldDelta.pf.
- FieldBundleDelta only partially completed.

Intermediate progress on FieldBudleDelta tests

Tests pass.

Still need more tests to check treatment of ungridded dims.

FieldDelta and FieldBundleDelta pass tests.

Have commented out logic for controlling time varying - saving for later feature.

## Types of change(s)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)
- [ ] Refactor (no functional changes, no api changes)

## Checklist
- [ ] Tested this change with a run of GEOSgcm
- [x] Ran the Unit Tests (`make tests`)

### CODE REVIEW

File: esmf_utils/UngriddedDims.F90 [hunk 1]
```diff
@@ -11,6 +11,7 @@ module mapl3g_UngriddedDims
    use esmf, only: ESMF_InfoCreate
    use esmf, only: ESMF_InfoSet
    use esmf, only: ESMF_InfoDestroy
+   use esmf, only: ESMF_InfoPrint
```
Comment (darianboggs) [inline esmf_utils/UngriddedDims.F90:14]: This may be unused.

Comment (tclune) [reply esmf_utils/UngriddedDims.F90:14]:
```suggestion
```

File: field_utils/FieldDelta.F90 [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+      subroutine update_num_levels(num_levels, field, ignore, rc)
+         integer, optional, intent(in) :: num_levels
+         type(ESMF_Field), intent(inout) :: field
+         character(*), intent(in), optional :: ignore
+         integer, optional, intent(out) :: rc
+
+         integer :: status
+
+         _RETURN_UNLESS(present(num_levels))
+         _RETURN_IF(ignore == 'num_levels')
```
Comment (darianboggs) [inline field_utils/FieldDelta.F90:237]: Does `_RETURN_IF` check if `ignore` is present?

File: field_utils/FieldDelta.F90 [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+      subroutine update_units(units, field, ignore, rc)
+         character(*), optional, intent(in) :: units
+         type(ESMF_Field), intent(inout) :: field
+         character(*), intent(in), optional :: ignore
+         integer, optional, intent(out) :: rc
+
+         integer :: status
+
+         _RETURN_UNLESS(present(units))
+         _RETURN_IF(ignore == 'units')
```
Comment (darianboggs) [inline field_utils/FieldDelta.F90:253]: See comment in `update_num_levels` regarding `_RETURN_IF`.

File: field_utils/FieldBundleDelta.F90 [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+      ! unique attribute in bundle
+      call update_interpolation_weights(this%interpolation_weights, bundle, ignore=ignore_, _RC)
+
+      _RETURN(_SUCCESS)
+   contains
+
+      subroutine update_units(units, field, ignore, rc)
+         character(*), optional, intent(in) :: units
+         type(ESMF_Field), intent(inout) :: field
+         character(*), intent(in), optional :: ignore
```
Comment (tclune) [inline field_utils/FieldBundleDelta.F90:171]:
```suggestion
         character(*), intent(in) :: ignore
```

Comment (tclune) [reply field_utils/FieldBundleDelta.F90:171]: Outer procedure constructs a local variable that is always "present", so "optional" is misleading here.

File: field_utils/FieldDelta.F90 [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+      call update_num_levels(this%num_levels, field, ignore=ignore, _RC)
+      call update_units(this%units, field, ignore=ignore, _RC)
+
+      _RETURN(_SUCCESS)
+   contains
+
+      subroutine update_num_levels(num_levels, field, ignore, rc)
+         integer, optional, intent(in) :: num_levels
+         type(ESMF_Field), intent(inout) :: field
+         character(*), intent(in), optional :: ignore
```
Comment (tclune) [inline field_utils/FieldDelta.F90:231]:
```suggestion
         character(*), intent(in) :: ignore
```

File: field_utils/FieldDelta.F90 [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+      type(ESMF_Field), intent(inout) :: field
+      character(*), intent(in), optional :: ignore
+      integer, optional, intent(out) :: rc
+
+      integer :: status
+
+      call this%reallocate_field(field, ignore=ignore, _RC)
+
+      call update_num_levels(this%num_levels, field, ignore=ignore, _RC)
+      call update_units(this%units, field, ignore=ignore, _RC)
```
Comment (tclune) [inline field_utils/FieldDelta.F90:223]:
```suggestion
      call update_num_levels(this%num_levels, field, ignore=ignore_, _RC)
      call update_units(this%units, field, ignore=ignore_, _RC)
```

File: field_utils/FieldDelta.F90 (hunk 5 above)
Comment (tclune) [inline field_utils/FieldDelta.F90:231]:
```suggestion
         character(*), intent(in) :: ignore
```

File: field_utils/FieldDelta.F90 [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+
+         call MAPL_InfoSetInternal(field, key=KEY_NUM_LEVELS, value=num_levels, _RC)
+
+         _RETURN(_SUCCESS)
+      end subroutine update_num_levels
+
+      subroutine update_units(units, field, ignore, rc)
+         character(*), optional, intent(in) :: units
+         type(ESMF_Field), intent(inout) :: field
+         character(*), intent(in), optional :: ignore
```
Comment (tclune) [inline field_utils/FieldDelta.F90:247]:
```suggestion
         character(*), intent(in) :: ignore
```

File: field_utils/FieldDelta.F90 [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+      do i = 1, size(fieldList)
+         call this%reallocate_field(fieldList(i), ignore, _RC)
+      end do
+
+      _RETURN(_SUCCESS)
+   end subroutine reallocate_fields
+
+   ! TODO - delete when next ESMF release provides support.
+   
+   impure elemental logical function ESMF_GeomEqual(geom1, geom2)
```
Comment (darianboggs) [inline field_utils/FieldDelta.F90:429]: This seems pure since geom1 and geom2 are intent(in).

File: field_utils/FieldDelta.F90 [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+         call ESMF_GeomGet(geom1, xgrid=xgrid1)
+         call ESMF_GeomGet(geom2, xgrid=xgrid2)
+         ESMF_GeomEqual = (xgrid1 == xgrid2)
+         return
+      end if
+      
+   end function ESMF_GeomEqual
+
+
+   impure elemental logical function ESMF_GeomNotEqual(geom1, geom2)
```
Comment (darianboggs) [inline field_utils/FieldDelta.F90:476]: This seems pure since geom1 and geom2 are intent(in).

File: field_utils/FieldUtilities.F90 [hunk 10]
```diff
@@ -303,116 +204,127 @@ subroutine FieldPow(field_out,field_in,expo,rc)
...
 
-      _RETURN(_SUCCESS)
-         
-   end subroutine FieldUpdate_from_attributes
+         integer :: status
+         type(ESMF_GeomType_Flag) :: geomtype
+         type(ESMF_Grid) :: grid
 
+         call ESMF_FieldBundleGet(fieldBundle, geomtype=geomtype, _RC)
+         if (geomtype == ESMF_GEOMTYPE_GRID) then
```
Comment (darianboggs) [inline field_utils/FieldUtilities.F90:291]: This looks like the geom for a FieldBundle can only be a grid. Is it a MAPL constraint that FieldBundle object can only have a Grid geometry?

File: field_utils/FieldBundleDelta.F90 [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+      associate (field_delta => FieldDelta(geom=geom, typekind=typekind, num_levels=num_levels, units=units))
+        bundle_delta = FieldBundleDelta(field_delta, fieldCount, interpolation_weights)
+      end associate
+
+   end function new_FieldBundleDelta
+
+   function new_FieldBundleDelta_field_delta(field_delta, fieldCount, interpolation_weights) result(bundle_delta)
+      type(FieldBundleDelta) :: bundle_delta
+      type(FieldDelta), intent(in) :: field_delta
+      integer, optional, intent(in) :: fieldCount
```
Comment (darianboggs) [inline field_utils/FieldBundleDelta.F90:58]: Does this get used?

File: field_utils/FieldBundleDelta.F90 [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+      call MAPL_FieldBundleGet(bundle, fieldList=fieldList, _RC)
+      call this%field_delta%update_fields(fieldList, ignore=ignore_, _RC)
+
+      ! unique attribute in bundle
+      call update_interpolation_weights(this%interpolation_weights, bundle, ignore=ignore_, _RC)
+
+      _RETURN(_SUCCESS)
+   contains
+
+      subroutine update_units(units, field, ignore, rc)
```
Comment (darianboggs) [inline field_utils/FieldBundleDelta.F90:168]: Does this get used?

File: field_utils/FieldDelta.F90 (hunk 9 above)
Comment (tclune) [reply field_utils/FieldDelta.F90:476]: The ESMF calls are not themselves PURE.  So ...

File: field_utils/FieldBundleDelta.F90 (hunk 11 above)
Comment (tclune) [reply field_utils/FieldBundleDelta.F90:58]: Probably not yet.    It is just a wrapper for the full constructor (or vice versa).  When deployed in couplers we can circle back to see which constructors here should survive.

File: field_utils/FieldBundleDelta.F90 (hunk 12 above)
Comment (tclune) [reply field_utils/FieldBundleDelta.F90:168]: Ah good catch.  Holdover from initial implementation.

File: field_utils/FieldBundleDelta.F90 [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+         _RETURN_UNLESS(present(units))
+         _RETURN_IF(ignore == 'units')
+
+         call MAPL_FieldBundleGet(bundle, fieldList=fieldList, _RC)
+         do i = 1, size(fieldList)
+            call MAPL_InfoSetInternal(fieldList(i), key=KEY_UNITS, value=units, _RC)
+         end do
+
+         _RETURN(_SUCCESS)
+      end subroutine update_units
```
Comment (tclune) [inline field_utils/FieldBundleDelta.F90:187]:
```suggestion
```

File: field_utils/FieldBundleDelta.F90 [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+      type(ESMF_Info) :: vertical_info
+      integer :: old_field_count, new_field_count
+      integer :: num_levels
+      character(:), allocatable :: units, vloc
+      character(ESMF_MAXSTR), allocatable :: fieldNameList(:)
+
+      ! Easy case 1: field count unchanged
+      call MAPL_FieldBundleGet(bundle, fieldList=fieldList, _RC)
+      _RETURN_UNLESS(allocated(this%interpolation_weights))
+      new_field_count = size(this%interpolation_weights) - 1
```
Comment (darianboggs) [inline field_utils/FieldBundleDelta.F90:217]: Why is the number of new fields one less than the number of interpolation_weights?

File: field_utils/FieldUtilities.F90 (hunk 10 above)
Comment (tclune) [reply field_utils/FieldUtilities.F90:291]: I just did not want to waste time on nonexistent use cases just yet.   I've got a PR into ESMF to generalize this  (geom optional argument) which will make the logic here much simpler.

File: field_utils/FieldBundleDelta.F90 (hunk 14 above)
Comment (tclune) [reply field_utils/FieldBundleDelta.F90:217]: The first element of the weights is a constant offset.   This is to handle the case when the rule is to set an ext data export to a constant value.   Easiest way to handle this was just to represent with empty bundle.  I will add a comment in source.

File: field_utils/tests/Test_FieldBundleDelta.pf [hunk 15]
```diff
@@ -0,0 +1,530 @@
...
+      call MAPL_FieldBundleGet(bundle, fieldList=fieldList)
+      do i = 1, size(fieldList)
+         call ESMF_FieldDestroy(fieldList(i))
+      end do
+      call ESMF_FieldBundleDestroy(bundle)
+
+   end subroutine teardown_bundle
+
+   @test(type=ESMF_TestMethod, npes=[1])
+   subroutine test_change_typekind(this)
```
Comment (darianboggs) [inline field_utils/tests/Test_FieldBundleDelta.pf:156]: Does this check if the typekind has changed in the bundle?

File: field_utils/FieldBundleDelta.F90 (hunk 14 above)
Comment (tclune) [inline field_utils/FieldBundleDelta.F90:217]:
```suggestion
      ! The number of weights is always one larger than the number of fields to support a constant
      ! offset.  ("Weights" is a funny term in that case.)
      new_field_count = size(this%interpolation_weights) - 1
```

File: generic3g/couplers/CouplerMetaComponent.F90 [hunk 16]
```diff
@@ -101,6 +101,27 @@ recursive subroutine initialize(this, importState, exportState, clock, rc)
       _RETURN(_SUCCESS)
    end subroutine initialize
 
+   ! Check if export item has been updated and update import item
+   ! accordingly.
+   recursive subroutine update_time_varying(this, importState, exportState, clock, rc)
```
Comment (darianboggs) [inline generic3g/couplers/CouplerMetaComponent.F90:106]: It looks like this does nothing?

File: generic3g/couplers/CouplerMetaComponent.F90 [hunk 17]
```diff
@@ -101,6 +101,27 @@ recursive subroutine initialize(this, importState, exportState, clock, rc)
...
+      integer, optional, intent(out) :: rc
+
+      integer :: status
+      type(ESMF_Field) :: f_in, f_out
+
+!#      _RETURN_UNLESS(this%export_is_time_varying())
+      call ESMF_StateGet(importState, itemName='import[1]', field=f_in, _RC)
+      call ESMF_StateGet(exportState, itemName='export[1]', field=f_out, _RC)
+
+!#      call FieldUpdate(f_in, from=f_out, ignore=this%action%get_ignore(), _RC)
```
Comment (darianboggs) [inline generic3g/couplers/CouplerMetaComponent.F90:120]: Is this a work in progress?

**Review by darianboggs — COMMENTED** [review]
I have reviewed all the files.

File: field_utils/FieldDelta.F90 (hunk 8 above)
Comment (darianboggs) [reply field_utils/FieldDelta.F90:429]: Perhaps they are supposed to be intent(inout).

File: field_utils/tests/Test_FieldBundleDelta.pf (hunk 15 above)
Comment (tclune) [reply field_utils/tests/Test_FieldBundleDelta.pf:156]: Yes - albeit indirectly by retrieving a pointer to an R8 array.

File: field_utils/tests/Test_FieldBundleDelta.pf [hunk 18]
```diff
@@ -0,0 +1,530 @@
...
+
+      call MAPL_FieldBundleGet(bundle, fieldList=fieldList, _RC)
+      @assert_that(size(fieldList), is(FIELD_COUNT))
+
+      do i = 1, FIELD_COUNT
+         call ESMF_FieldGet(fieldList(i), fArrayPtr=x_r4, _RC)
+         @assert_that(shape(x_r4), is(equal_to([6,6])))
+
+         call MAPL_InfoGetInternal(fieldList(i), KEY_UNITS, value=new_units, _RC)
+         @assertEqual('km', new_units)
```
[the remark this answers is not in the document]
Comment (tclune) [reply field_utils/tests/Test_FieldBundleDelta.pf:256]: No.  It is confirming that they did not change.   Delta only specified a new geom.

**Comment by darianboggs** [comment]
I'll take a look at this in the morning when I am more fresh.

**Comment by darianboggs** [comment]
My latest thoughts, @tclune:

We could store the keys as named constants in the MAPL _type modules_ to which they correspond. Those keys would be used to make/set ESMF_Info and get ESMF_Info. It would be hiearchical like ESMF_Info. An object would contain only the keys for values at it's own level of the hierarchy.

It could be bound to the type, but I don't think it needs to be.

This would help in our development because we could find the keys if we know which object we are dealing with, and they would named, so we would not need to use string literals.

Beyond that, what you have makes sense.

What is the distinction between Internal, Private, Shared, and normal keys? I've forgotten.

Let me know if and when you want to discuss it further.

**Comment by tclune** [comment]
> What is the distinction between Internal, Private, Shared, and normal keys? I've forgotte

* PRIVATE:  user-level info objects specific to a component
* SHARED: user-level info objects that must be consistent across connections (couplers must copy along the chain)
* INTERN: MAPL internal use only

**Comment by tclune** [comment]
@darianboggs   Please re-review.   I had not gotten around to fixing a couple of minor issues until over the weekend.

**Review by darianboggs — DISMISSED** [review]
Looks good

**Pull request merged**